APE_Black_Formatting.rst

```diff
@@ -0,0 +1,286 @@
+this APE's conventions in the future, this APE is concerned solely with the core
+package itself.
+
+Code formatting is controversial and the `astropy-dev thread`_ about Black
+generated a spirited discussion but did not result in a clear consensus. This
+strong division of opinions about automated code formatting is common and has
+been documented in other projects, for instance in `Django DEP-0008`_.
+
+We do not expect unanimous agreement on this APE but do hope to demonstrate the
+value of Black for the collective needs of the community and Astropy Project.
+
+.. _astropy-dev thread: https://groups.google.com/g/astropy-dev/c/6cRJCMgaFyM/
+
+Note that ideas and some text for this APE have been adapted from a similar
+enhancement proposal for Django `Django DEP-0008`_.
+
+Rationale
+^^^^^^^^^
+
+**Consistent and readable code formatting** reduces effort in both code review
+and in future maintenance. Code is read far more often than it is written, so it
+makes sense to invest effort to ensure that the code is easy to read later.
+
+Ensuring that code formatting meets ``astropy`` standards is currently done by
+applying CI checks based on :pep:`8`. Learning and consistently applying all
+these rules is not easy. There are at least 200 commits in ``astropy`` which
+were solely address PEP8 violations, so this is a real issue.
+
+Because of the need to follow these format guidelines, ``astropy`` developers
+must pay close attention to formatting, constantly making small decisions about
+code formatting based on rules, on taste, or on a mix of both. While some of
+:pep:`8` is focused on guidelines like naming conventions that cannot be
+automated, much of it is essentially algorithmic and hence these decisions are a
+needless burden on the developers.
+
+With **automated code formatting**, the workload of code formatting is done by
+computers, saving human bandwidth for higher value and complexity activities
+like the logic or interface design. Collectively, contributors no longer need to
+think about code formatting when writing or in code reviews. By ceding control
+to Black_ they can focus on the content of the code and not worry about details
+like where to break a line or how to arrange a list.  Whatever the formatter
+does is the expected result, by definition.
+
+Automated code formatting will increase consistency across the ``astropy`` code
+base. Currently the style of each subpackage reflects the preferences of the
+maintainers. For instance, the ``time``, ``table``, and ``io.ascii`` subpackages
+allow line lengths of up to 100, while most others use 80 or 88. There are other
+more subtle differences and astropy contributors find this confusing. This has
+come up in the past.
+
+While there are multiple code formatters for Python, the Black Python code
+formatter produces good enough results that it is getting significant traction
+in many open source projects, including: scikit-learn, pytest, tox, Pyramid,
```

Comment:
for completeness I would also suggest to cite the  projects that decide of not to go down this route. E.g. quite a couple of projects are waiting for black to implement some reasonable approach for mathematical and nested structure formatting. 

For an absolutely ugly reformatting example, one can look at scipy: https://github.com/scipy/scipy/pull/14330 where black wouldn't result a more readable code, but a bloat of 45% of number of lines.

Reply:
@bsipocz - good idea, can you provide a list of other projects that have decided to either (1) reject auto-formatting entirely as being a bad idea or (2) wait until the black implementation is better.

The APE states that cases where black does poorly can and should be excepted. Based on testing by myself for time/table/io.ascii and @eteq (coordinates), requiring exceptions is not common but they will be allowed. Any others e.g. @WilliamJamieson @nstarman have data on the need for exceptions?

Reply:
a few projects (scipy is rejected, numpy is waiting, they may be others, too) are waiting for the math formatting summit, and the promise from the side of the black developers to take scientific python needs into account. Most of those mailing list and discourse and PR discussions are linked to the scipy and numpy PRs cited above.

afaik all that implemented it and likes it, in fact likes the fact that linting is now enforced, rather than the formatting itself, and thus my opinion of enforcing linting would do the job without enforcing to live together with the nonsense of the choices black forces on us (e.g. all of those in the project were against linting cited less serious issues and have been unwilling to accept the choices of other team members, but now are willing to accept the nonesences around the 3rd party, black's verbose whitespacing?)

Reply:
@bsipocz -I don't know precisely what you mean by "enforcing linting". We already do CI flake8 checks, so do you mean something different?

What I like most about black is fast *auto-formatting* in my IDE. But in general I am happy with the format choices black makes, so you should not presume to speak to the opinions of the large community of black users.

Reply:
currently, a select number of linting options are checked, rather than only a few ignored, and that is historical reasons as there was a push against being strict on linting, being conservative and allowing interpretation of the pep8 rules as basically, the aim was to keep the heterogeneity of the different submodules. We wanted and could have done e.g. autopep8 to auto format those things that are uncontroversial. 
So now, the proposal is to be being instead extra strict and enforcing linting choices that are blindly doing PEP8 and not necessarily ideal for the codebase we have. 

And I don't claim to know the opinion of all black users, but say that there is a well documented shared concern in the wider scientific python community that its enforced code style is not ideal for math and array-heavy usage. Beyond this shared concerns I have personal issues with the superfluous white spaces it applies, the 30-45% addition of lines with no useful information (single parens or characters).  I also have personal preferences for IDE and editor behaviour which I'm very happy with, but am not trying to force that on all the other developers.

Reply:
> The APE states that cases where black does poorly can and should be excepted. Based on testing by myself for time/table/io.ascii and @eteq (coordinates), requiring exceptions is not common but they will be allowed. Any others e.g. @WilliamJamieson @nstarman have data on the need for exceptions?

In my experience, I haven't found too many places where exceptions are needed.
Sometimes / often a blackened math one-liner looks worse because of the spaces around operators. However, I tend to have comments by math code with a cleaner representation of the math, because except for very simple expressions, I don't find math in **Python** very readable, regardless of style. For multi-line complex math, my thoughts on Python and needed explanatory comments are even more true. IMO Black doesn't usually make the **end** result worse as it encourages me to rewrite the math.

tl;dr, piles of multi-line math are much less readable for future maintainers than well factored code with comments. Sometimes multi-line math is unavoidable, which is why we can selectively turn black off.